src/Relay/createEnvironment.ts

```diff
@@ -1,15 +1,91 @@
-import { Environment, Network, RecordSource, Store } from "relay-runtime"
-import { metaphysics } from "../Utils/metaphysics"
+import "isomorphic-fetch"
+import "regenerator-runtime/runtime"
 
-export function createEnvironment(user?: User) {
-  const fetchQuery = (operation, variables, cacheConfig, uploadables) => {
-    return metaphysics({ query: operation.text, variables }, user)
+import RelayServerSSR from "react-relay-network-modern-ssr/lib/server"
+import RelayClientSSR from "react-relay-network-modern-ssr/lib/client"
+import { Environment, RecordSource, Store } from "relay-runtime"
+import { NetworkError } from "../Utils/errors"
+import { data as sd } from "sharify"
+
+import {
+  RelayNetworkLayer,
+  urlMiddleware,
+  cacheMiddleware,
+  loggerMiddleware,
+} from "react-relay-network-modern"
+
+interface Config {
+  cache?: object
+  user?: User
+  checkStatus?: boolean
+}
+
+interface RelayEnvironment extends Environment {
+  relaySSRMiddleware: RelayClientSSR | RelayServerSSR
+}
+
+export function createEnvironment(config: Config = {}) {
+  const { cache = {}, checkStatus, user } = config
+  const isServer = typeof window === "undefined"
+  const relaySSRMiddleware = isServer
+    ? new RelayServerSSR()
+    : new RelayClientSSR(cache)
+
+  relaySSRMiddleware.debug = false
+
+  const headers = {
+    "Content-Type": "application/json",
+    "User-Agent": "Reaction",
   }
-  const network = Network.create(fetchQuery)
+
+  const url = isServer
+    ? process.env.METAPHYSICS_ENDPOINT
+    : sd.METAPHYSICS_ENDPOINT
+
+  const network = new RelayNetworkLayer([
+    urlMiddleware({
+      url,
+      headers: !!user
+        ? {
+            ...headers,
+            "X-USER-ID": user && user.id,
+            "X-ACCESS-TOKEN": user && user.accessToken,
+          }
+        : headers,
+    }),
+    relaySSRMiddleware.getMiddleware({
+      lookup: true,
+    }),
+    cacheMiddleware({
+      size: 100, // max 100 requests
+      ttl: 900000, // 15 minutes
```

Comment:
The TTL could probably be increased greatly, although I guess it doesn’t persist and so a user’s session doesn’t last much longer? FYI While we may still need to fine-tune it, for now we have the response caching TTL in Emission set to a day.